File: pocs/integrations/src/graph/stub.ts
import type { GraphClient } from "./types.js";

export function createStubGraphClient(): GraphClient {
  async function upsertEntity(op: Parameters<GraphClient["upsertEntity"]>[0]): Promise<void> {
    const propCount = Object.keys(op.properties).length;
    console.log(`[graph] UPSERT ${short(op.entityType)} id=${op.entityId} (${propCount} props)`);
    for (const [url, val] of Object.entries(op.properties)) {
      console.log(`  ${short(url)} = ${JSON.stringify(val)}`);
    }
    if (op.provenance.location?.name) console.log(`  source: ${op.provenance.location.name}`);
  }

  async function upsertLink(op: Parameters<GraphClient["upsertLink"]>[0]): Promise<"ok"> {
    console.log(`[graph] LINK ${short(op.linkType)} ${op.sourceEntityId} -> ${short(op.targetEntityType)} id=${op.targetId}`);
    return "ok";
  }

  return {
    upsertEntity,
    async bulkUpsertEntities(ops) {
      const start = Date.now();
      const ok: string[] = [];
      for (const op of ops) { await upsertEntity(op); ok.push(String(op.entityId)); }
      return { ok, failed: [], batches: 1, fellBackBatches: 0, durationMs: Date.now() - start };
    },
    upsertLink,
    async bulkUpsertLinks(ops, opts) {
      const start = Date.now();
      const ok: string[] = [];
      for (const op of ops) {
        await upsertLink(op);
        ok.push(op.opId);
      }
      if (opts?.onBatchOk) await opts.onBatchOk(ok);
      return { ok, failed: [], batches: 1, fellBackBatches: 0, durationMs: Date.now() - start };
    },
    async archiveEntity(op) {
      console.log(`[graph] ARCHIVE ${short(op.entityType)} id=${op.entityId}`);
    },
    identity: () => "stub:local",
    async hasEntity() {
      return true;
    },
  };
}

function short(url: string): string {
  return url.split("/entity-type/")[1] ?? url.split("/property-type/")[1] ?? url;
}

File: pocs/integrations_rs/src/graph/stub.rs
//! Graphless client: logs every op, accepts everything. `has_entity` answers
//! true so adopted state passes the sentinel probe against the stub.

use error_stack::Report;

use crate::error::GraphError;
use crate::value::js_string;

use super::{ArchiveOp, BatchOk, BulkResult, EntityOp, GraphClient, LinkOp};

pub struct StubClient;

fn short(type_url: &str) -> &str {
    type_url.split("/entity-type/").nth(1).unwrap_or(type_url)
}

#[async_trait::async_trait]
impl GraphClient for StubClient {
    fn identity(&self) -> String {
        "stub:graph".to_owned()
    }

    async fn has_entity(&self, _full_entity_id: &str) -> Result<bool, Report<GraphError>> {
        Ok(true)
    }

    async fn archive_entity(&self, op: &ArchiveOp) -> Result<(), Report<GraphError>> {
        tracing::info!(
            "[graph] ARCHIVE {} id={}",
            short(&op.entity_type),
            op.entity_id
        );
        Ok(())
    }

    async fn bulk_upsert_entities(&self, ops: Vec<EntityOp>, on_batch_ok: BatchOk) -> BulkResult {
        let mut ok = vec![];
        for op in &ops {
            tracing::info!(
                "[graph] UPSERT {} id={} ({} props)",
                short(&op.entity_type),
                js_string(&op.entity_id),
                op.properties.len()
            );
            ok.push(js_string(&op.entity_id));
        }
        on_batch_ok(ok.clone()).await;
        BulkResult {
            ok,
            ..BulkResult::default()
        }
    }

    async fn bulk_upsert_links(&self, ops: Vec<LinkOp>, on_batch_ok: BatchOk) -> BulkResult {
        let mut ok = vec![];
        for op in &ops {
            tracing::info!(
                "[graph] LINK {} {} -> {}",
                short(&op.link_type),
                op.source_entity_id,
                op.target_id
            );
            ok.push(op.op_id.clone());
        }
        on_batch_ok(ok.clone()).await;
        BulkResult {
            ok,
            ..BulkResult::default()
        }
    }
}
